### PR TITLE
feat: add wait operation

### PR DESCRIPTION
**Pull request by shubham-cmyk**
## Explanation

Add wait operation similar to describe 

## Related issue
Fixes : https://github.com/kyverno/chainsaw/issues/888

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.

### CODE REVIEW

File: pkg/apis/v1alpha1/wait.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	// ObjectLabelsSelector determines the selection process of referenced objects.
+	ObjectLabelsSelector `json:",inline"`
+
+	// Condition represents the specific condition to wait for.
+	// Example: "Available", "Ready", etc.
+	Condition string `json:"condition"`
+
+	// AllNamespaces indicates whether to wait for resources in all namespaces.
+	// +optional
+	AllNamespaces bool `json:"allNamespaces,omitempty"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:27]: We don't have that option for other collectors

File: pkg/apis/v1alpha1/wait.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+	// Resource type on which the wait operation will be applied.
+	Resource string `json:"resource"`
+
+	// ObjectLabelsSelector determines the selection process of referenced objects.
+	ObjectLabelsSelector `json:",inline"`
+
+	// Condition represents the specific condition to wait for.
+	// Example: "Available", "Ready", etc.
+	Condition string `json:"condition"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:23]: Wait works with delete or conditions, i should put it inside a struct

Comment (eddycharly) [reply pkg/apis/v1alpha1/wait.go:23]:
```
 Alternatively, the command can wait for the given set of resources to be deleted by providing the "delete" keyword as
the value to the --for flag.
```

**Comment by eddycharly** [comment]
@shubham-cmyk i resolved conflicts and added some comments

**Comment by shubham-cmyk** [comment]
@eddycharly I am considering 

```
WaitType WaitType `json:"waitType"`

```

```
// WaitType represents the type of wait operation.
type WaitType string

// Define constants for WaitType values
const (
	WaitTypeDelete    WaitType = "delete"
	WaitTypeCondition WaitType = "condition"
)
```

File: pkg/apis/v1alpha1/wait.go (hunk 1 above)
Comment (shubham-cmyk) [reply pkg/apis/v1alpha1/wait.go:41]: @eddycharly I didn't get that should I remove this ?

**Comment by eddycharly** [comment]
@shubham-cmyk no need for a `WaitType` IMHO, just create separate structs.

**Comment by eddycharly** [comment]
Something like:
```golang
type For Deletion {
}

type For Condition {
}

type For struct {
  Deletion *Deletion
  Condition *Condition
}
```

File: pkg/apis/v1alpha1/wait.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type For struct {
+	// +optional
+	Deletion *Deletion `json:"delete,omitempty"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:7]: Needs comment

File: pkg/apis/v1alpha1/wait.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	// +optional
+	Condition *Condition `json:"condition,omitempty"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:9]: Needs comment

File: pkg/apis/v1alpha1/wait.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// The specific condition to wait for, e.g., "Available", "Ready".
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:17]:
```suggestion
	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
```

File: pkg/validation/test/wait.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		}
+		if obj.Name != "" && obj.Selector != "" {
+			errs = append(errs, field.Invalid(path, obj, "a name or label selector must be specified (found both)"))
+		}
+		errs = append(errs, ValidateFor(path.Child("for"), &obj.For)...)
+	}
+	return errs
+}
+
+func ValidateFor(path *field.Path, obj *v1alpha1.For) field.ErrorList {
```
Comment (eddycharly) [inline pkg/validation/test/wait.go:22]: Should go in `for.go` ;)

File: pkg/runner/kubectl/wait.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+	if collector.Name != "" {
+		args = append(args, fmt.Sprintf("%s/%s", collector.Resource, collector.Name))
+	} else {
+		args = append(args, collector.Resource)
+	}
+
+	if collector.For.Deletion != nil {
+		args = append(args, "--for=delete")
+	} else if collector.For.Condition != nil && collector.For.Condition.ConditioName != "" {
```
Comment (eddycharly) [inline pkg/runner/kubectl/wait.go:31]: `collector.For.Condition.ConditioName != ""` should be in an inner if statement

Comment (eddycharly) [reply pkg/runner/kubectl/wait.go:31]: Also it should return a dedicated error

File: pkg/runner/kubectl/wait.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	if collector.Namespace != "" {
+		args = append(args, "-n", collector.Namespace)
+	} else {
+		args = append(args, "-n", "$NAMESPACE")
+	}
+	if collector.Timeout != nil {
+		args = append(args, fmt.Sprintf("--timeout=%s", collector.Timeout.Duration.String()))
+	}
```
Comment (eddycharly) [inline pkg/runner/kubectl/wait.go:48]: I think we need to disable the timeout in the else.

Comment (shubham-cmyk) [reply pkg/runner/kubectl/wait.go:48]: since sleep infinity works i guess it would work ?

File: pkg/apis/v1alpha1/wait.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// OutputFormatType defines the enum for valid output formats.
+type OutputFormatType string
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:6]: We already have something defined for that (see in other collectors), we only support yaml and json for now.

File: pkg/apis/v1alpha1/wait.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
+	ConditionName string `json:"conditionName"`
+	// The specific condition status to wait for, e.g., "True", "False".
+	// +optional
+	ConditionValue *bool `json:"conditionValue,omitempty"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:40]:
```suggestion
	ConditionValue string `json:"conditionValue,omitempty"`
```

File: pkg/apis/v1alpha1/wait.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+	Condition *Condition `json:"condition,omitempty"`
+}
+
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
+	ConditionName string `json:"conditionName"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:37]: We can use `name` and `value`, no need to prefix with `condition`

File: pkg/apis/v1alpha1/wait.go (hunk 10 above)
Comment (shubham-cmyk) [reply pkg/apis/v1alpha1/wait.go:40]: why not bool ?

File: pkg/runner/kubectl/wait.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package kubectl
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/kyverno/chainsaw/pkg/apis/v1alpha1"
+)
+
+func WaitForResource(collector *v1alpha1.Wait) (*v1alpha1.Command, error) {
```
Comment (eddycharly) [inline pkg/runner/kubectl/wait.go:10]:
```suggestion
func Wait(collector *v1alpha1.Wait) (*v1alpha1.Command, error) {
```

File: pkg/runner/kubectl/wait.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+	if collector.Namespace != "" {
+		args = append(args, "-n", collector.Namespace)
+	} else {
+		args = append(args, "-n", "$NAMESPACE")
+	}
+	if collector.Timeout != nil {
+		args = append(args, fmt.Sprintf("--timeout=%s", collector.Timeout.Duration.String()))
+	} else {
+		args = append(args, fmt.Sprintf("--timeout=%s", "infinity"))
```
Comment (eddycharly) [inline pkg/runner/kubectl/wait.go:56]:
```
k wait pods --for=condition=Ready="true" -n local-path-storage --all --timeout=infinity
error: invalid argument "infinity" for "--timeout" flag: time: invalid duration "infinity"
See 'kubectl wait --help' for usage.
```

Comment (eddycharly) [reply pkg/runner/kubectl/wait.go:56]: We should have an end to end test for that (in `/testdata/e2e/examples`

File: pkg/apis/v1alpha1/wait.go (hunk 10 above)
Comment (eddycharly) [reply pkg/apis/v1alpha1/wait.go]: Because `kubectl wait ...` expects a string

Comment (eddycharly) [reply pkg/apis/v1alpha1/wait.go]: `k wait pods --for=condition=Ready="something" -n local-path-storage --all`

Comment (shubham-cmyk) [reply pkg/apis/v1alpha1/wait.go]: i manage it from here  

```
args = append(args, fmt.Sprintf("--for=condition=%s=%t", collector.For.Condition.ConditionName, *collector.For.Condition.ConditionValue))
```

File: pkg/apis/v1alpha1/format.go [hunk 14]
```diff
@@ -1,5 +1,5 @@
 package v1alpha1
 
 // Format determines the output format (json or yaml).
-// +kubebuilder:validation:Enum:=json;yaml;
+// +kubebuilder:validation:Enum=json;yaml;name;go-template;go-template-file;template;templatefile;jsonpath;jsonpath-as-json;jsonpath-file
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/format.go:4]: We don't support all formats

File: pkg/apis/v1alpha1/wait.go (hunk 10 above)
Comment (eddycharly) [reply pkg/apis/v1alpha1/wait.go]: yeah but with a bool `Not yet` is not going to be accepted

File: pkg/apis/v1alpha1/format.go (hunk 14 above)
Comment (shubham-cmyk) [reply pkg/apis/v1alpha1/format.go:4]: :(

Comment (eddycharly) [reply pkg/apis/v1alpha1/format.go]: We can create a tracking issue but we don't really have use case for more complex formats today so 🤷

File: pkg/apis/v1alpha1/wait.go [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+	Condition *Condition `json:"condition,omitempty"`
+}
+
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
+	ConditionName string `json:"name"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:21]:
```suggestion
	Name string `json:"name"`
```

File: pkg/apis/v1alpha1/wait.go [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+	// The specific condition status to wait for, e.g., "True", "False".
+	// +optional
+	ConditionValue string `json:"value,omitempty"`
```
[the remark this answers is not in the document]
Comment (eddycharly) [reply pkg/apis/v1alpha1/wait.go:24]: We need a pointer here, `nil` is not the same as `""`

File: pkg/validation/test/for.go [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+	var errs field.ErrorList
+	if obj != nil {
+		if obj.Deletion == nil && obj.Condition == nil {
+			errs = append(errs, field.Invalid(path, obj, "either a deletion or a condition must be specified"))
+		}
+		if obj.Deletion != nil && obj.Condition != nil {
+			errs = append(errs, field.Invalid(path, obj, "a deletion or a condition must be specified (found both)"))
+		}
+		if obj.Condition != nil && obj.Condition.ConditionName == "" {
+			errs = append(errs, field.Invalid(path.Child("conditionName"), obj, "a condition name must be specified"))
```
Comment (eddycharly) [inline pkg/validation/test/for.go:18]:
```suggestion
			errs = append(errs, field.Invalid(path.Child("name"), obj, "a condition name must be specified"))
```

File: pkg/apis/v1alpha1/wait.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
+	Name string `json:"name"`
+	// The specific condition status to wait for, e.g., "True", "False".
+	// +optional
+	Value string `json:"value,omitempty"`
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:24]:
```suggestion
	Value *string `json:"value,omitempty"`
```

File: pkg/apis/v1alpha1/wait.go [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+	// +optional
+	Condition *Condition `json:"condition,omitempty"`
+}
+
+// Deletion represents parameters for waiting on a resource's deletion.
+type Deletion struct{}
+
+// Condition represents parameters for waiting on a specific condition of a resource.
+type Condition struct {
+	// ConditioName is the specific condition to wait for, e.g., "Available", "Ready".
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:20]:
```suggestion
	// Name is the specific condition to wait for, e.g., "Available", "Ready".
```

File: pkg/apis/v1alpha1/wait.go [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+	Name string `json:"name"`
+	// The specific condition status to wait for, e.g., "True", "False".
```
Comment (eddycharly) [inline pkg/apis/v1alpha1/wait.go:22]:
```suggestion
	// Value defines the specific condition status to wait for, e.g., "True", "False".
```

File: pkg/validation/test/for.go [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+	var errs field.ErrorList
+	if obj != nil {
+		if obj.Deletion == nil && obj.Condition == nil {
+			errs = append(errs, field.Invalid(path, obj, "either a deletion or a condition must be specified"))
+		}
+		if obj.Deletion != nil && obj.Condition != nil {
+			errs = append(errs, field.Invalid(path, obj, "a deletion or a condition must be specified (found both)"))
+		}
+		if obj.Condition != nil && obj.Condition.Name == "" {
+			errs = append(errs, field.Invalid(path.Child("name"), obj, "a condition name must be specified"))
```
Comment (eddycharly) [inline pkg/validation/test/for.go:18]:
```suggestion
			errs = append(errs, field.Invalid(path.Child("condition").Child("name"), obj, "a condition name must be specified"))
```

**Review by eddycharly — APPROVED** [review]
LGTM

**Review by eddycharly — APPROVED** [review]

**Pull request merged**